Makefile

```diff
@@ -23,6 +23,19 @@ in: inplace # just a shortcut
 inplace:
 	$(PYTHON) setup.py build_ext -i
 
+meson:
+	perl -i -pe \
+		's@\[build-system\]@[build-system]\nbuild-backend = "mesonpy"\n@' \
+		pyproject.toml
+	pip install --editable .  --verbose --no-build-isolation \
+		--config-settings editable-verbose=true
+	perl -0777 -i -pe \
+		's@\[build-system\]\nbuild-backend = "mesonpy"\n@[build-system]@' \
+		pyproject.toml
```

Comment:
I would rather have a tiny helper scrip under `build_tools/` than add a dev dependency on perl.

Reply:
Also if you `ctrl-c` during the build, the second perl command is not called and then the next `make meson` call will add a second `build-backend = "meson"` line to `pyproject.toml` causing `pip/_vendor/tomli/_parser.py` to fail.

Better use a smarter helper script that does not add duplicated `build-backend` lines.

Reply:
I thought about this indeed :wink: but at the moment I see `make dev-meson` as only useful during the transition period during the time Meson is not our default build-backend. This would be a convenient way for people to test Meson in their daily routine and report issues they bumped into. Once Meson is our default build-backend we can remove `make dev-meson`

If we want to be serious about developer API, I think we should look at `spin` or Scipy dev.py (they are different for what seems historical reasons). Personally I think we should discuss about this but in a dedicated issue, in order to keep this PR about Meson support, which is already not that easy to review.

Also you only have to do `make dev-meson` once. After that `import sklearn` will recompile as needed, so I feel the trade-off between robustness and usefulness is OK enough for the transition period. Happy to hear other opinions.

Reply:
`spin` is meant to [take the concept of `dev.py` and make it customizable and accessible](https://github.com/scientific-python/spin?tab=readme-ov-file#history) to other projects. As such, it's easy to extend and to write custom commands for it. But it does essentially the same thing as dev.py, and comes bundled with a few standard commands typically used across the ecosystem. For examples of how to use and customize spin, you can take a look at numpy, scikit-image, and dipy (see their `.spin` directories).

Reply:
> Also you only have to do make dev-meson once. After that import sklearn will recompile as needed, so I feel the trade-off between robustness and usefulness is OK enough for the transition period. Happy to hear other opinions.

Unless you get an error the first time you run it which happens to be the most likely time to experience a build error because your forgot to install Cython or any other build dependency in a new conda env for instance.

Reply:
It has already happened to me 3 times when playing with this PR while swapping dev envs because I was concurrently reviewing other PRs.

Reply:
The fact that spin is also used by other meson-based scientific python projects makes it interesting to learn.

The GDB / LLDB debug commands might also be interesting as it's typically tedious to remember the commandline args to get a debugger to work on Cython based project.

Reply:
`spin` looks cool but let's work on a possible spin integration in a follow-up PR to avoid side-tracking this PR too much.

Reply:
fd2c2fd995b35ae32bfe8791764ae266461ee0f7 is enough to address my concerns with ctrl-c / or failing builds.